### PR TITLE
Add new image normalization classes

### PR DESCRIPTION
**Pull request by larrybradley**
This PR contains the following changes to the `visualization` subpackage:

* ~~Adds a `Norm` class to generate a `matplotlib.colors.Normalize` object from an interval and a stretch instance.~~
* ~~Adds a convenience `ImgNorm` class for quick interactive analysis (and used by the `fits2bitmap` command-line script).  `ImgNorm` replaces the `scale_image` function.~~
* Deprecates the `scale_image` function (in favor of the `ImgNorm` class).
* Adds a default stretch for the existing `Normalization` class.
* Adds default `vmin/vmax` for the `ManualInterval` class.
* Adds missing API docs for existing interval and stretch classes.
* Adds addition tests for the visualization subpackage.
* Augments and updates the visualization narrative docs.

Examples
----------

With the existing `ImageNormalize` class one needs to input the `vmin/vmax` values, which is a bit cumbersome, e.g.:

```
interval = MinMaxInterval()
vmin, vmax = interval.get_limits(image)
norm = ImageNormalize(vmin=vmin, vmax=vmax, stretch=SqrtStretch())
```

With the new ~~`Norm` class~~ inputs, one can simply do, e.g.:

```
norm = ImageNormalize(image, MinMaxInterval(), SqrtStretch())
```

And then the image is displayed with:
```
ax.imshow(image, origin='lower', norm=norm)
```

cc: @astrofrog

### CODE REVIEW

**Comment by astrofrog** [comment]
@larrybradley - I'm in favor of having conveniences such as these, but I think that having ``Normalize``, ``Norm``, and ``ImgNorm`` is confusing - instead I would propose making the two new classes you are adding here class methods for ``Normalize``. In fact, ``Norm`` could even be merged with the existing ``Normalize`` initializer so that you can either specify ``vmin`` and ``vmax`` or ``interval`` and ``image`` in ``__init__``.

``ImgNorm`` can then be made into a convenience class method (I'm trying to think of a good name).

File: astropy/visualization/stretch.py [hunk 1]
```diff
@@ -47,38 +47,36 @@ def _prepare(values, out=None, clip=True):
...
-    Base class for the stretch classes, which, when called with an array of
-    values in the range [0:1], return an transformed array of values, also in
-    the range [0:1].
+    Base class for the stretch classes, which, when called with an array
+    of values in the range [0:1], return an transformed array of values,
+    also in the range [0:1].
     """
 
-    def __call__(self, values, out=None, clip=True):
+    def __call__(self, values, clip=True, out=None):
```
Comment (astrofrog) [inline astropy/visualization/stretch.py]: Just out of curiosity, why the change?

**Comment by pllim** [comment]
> trying to think of a good name

`ConvenientNorm`? `NormForLazyPeople`? `EasyNorm`? `QuickNorm`?

File: astropy/visualization/stretch.py (hunk 1 above)
Comment (larrybradley) [inline astropy/visualization/stretch.py]: To make it consistent with the `BaseInterval` `__call__` method.  Also, the underlying `numpy` functions have `out` as the last keyword.

If you agree with `out` being last, then I should also change the calling sequence in the `stretch._prepare` function, so that everything is consistent.

**Comment by astrofrog** [comment]
@pllim - just to be clear, I meant a class method not class factory (so ``Normalize.from_...``)

**Comment by pllim** [comment]
`from_image`?

**Comment by larrybradley** [comment]
@astrofrog  Indeed, I was struggling with short-ish names (because this is a convenience class) similar to `ImageNormalize`.  Suggestions welcome!

**Comment by larrybradley** [comment]
@astrofrog  I assume you mean the `ImageNormalize` class (in astropy visualization), not the `Normalize` class (in matplotlib).

I don't think having these new classes as `ImageNormalize` methods is very convenient.  It's also very confusing.  Is something like this what you were thinking:

```
norm = ImageNormalize(stretch=SqrtStretch())
norm2 = norm.apply_interval(data, interval=MinMaxInterval())
imshow(data, norm=norm2)
```

I didn't want to change `ImageNormalize`, but it could be combined with `Norm`.  I think that is *much* better and less confusing than adding methods (and it requires one less unique class name :smile:).  It would require a bit of input validation checking to make sure the right combination of keywords are inputs.  Sound OK?

We'd still need a new name for `ImgNorm`.  This class essentially replaces the `scale_image` function, but `scale` can mean so many things, so it's not the best name.

**Comment by astrofrog** [comment]
> I didn't want to change ImageNormalize, but it could be combined with Norm. I think that is much better and less confusing than adding methods (and it requires one less unique class name 😄). It would require a bit of input validation checking to make sure the right combination of keywords are inputs. Sound OK?

Yep!

**Comment by larrybradley** [comment]
@astrofrog I've implemented the new changes:

* `ImageNormalize` can be initialized with either `data`/`interval` or `vmin`/`vmax` (which takes precedence).

* Replaced `scale_image` (deprecated) with a new function called `mpl_norm` that returns an `ImageNormalize` object.  (The `ImgNorm` class from the first implementation of this PR is gone).

File: CHANGES.rst [hunk 2]
```diff
@@ -29,9 +29,6 @@ New Features
 
 - ``astropy.modeling``
 
-  - Add a class to combine astropy fitters and functions to remove outliers
-    e. g., sigma clip. [#4760]
-
```
Comment (astrofrog) [inline CHANGES.rst]: Is there a reason why you are removing this?

File: CHANGES.rst [hunk 3]
```diff
@@ -160,24 +170,17 @@ Bug Fixes
...
-- ``astropy.coordinates``
-
-  - Initialization of ``Angle`` has been sped up for ``Quantity`` and ``Angle``
-    input. [#4970]
-
-  - The use of ``np.matrix`` instances in the transformations has been
-    deprecated, since this class does not allow stacks of matrices.  As a
-    result, the semi-public functions ``angles.rotation_matrix`` and
-    ``angles.angle_axis`` are also deprecated, in favour of the new routines
-    with the same name in ``coordinates.matrix_utilities``. [#5104]
```
Comment (astrofrog) [inline CHANGES.rst]: Same here - looks like an issue with conflict resolution?

**Comment by astrofrog** [comment]
@larrybradley - thanks for your work on this. I think ``simple_image_norm`` would be better than ``mpl_norm`` - changing ``mpl`` to ``image`` because it returns an ``ImageNormalization`` class, and adding ``simple`` to make it clearer that this is a convenience function that only supports a subset of cases.

**Comment by larrybradley** [comment]
@astrofrog Would `image_norm` be acceptable instead?  I understand the reasoning behind prepending `simple`, but that is explained in the docs.

File: CHANGES.rst (hunk 2 above)
Comment (larrybradley) [inline CHANGES.rst]: Not sure how that happened.  Thanks for catching it.  It's fixed.

File: CHANGES.rst (hunk 3 above)
Comment (larrybradley) [inline CHANGES.rst]: Fixed.

**Comment by larrybradley** [comment]
@astrofrog  I renamed the function `simple_norm`.

**Comment by astrofrog** [comment]
This all looks good to me now - thanks! Is this good to go from your side?

**Comment by larrybradley** [comment]
Yep, it's good to go.

File: astropy/visualization/interval.py [hunk 4]
```diff
@@ -6,39 +6,73 @@
 """
 
 from __future__ import division, print_function
-
```
Comment (MSeifert04) [inline astropy/visualization/interval.py]: why remove the blank line?

I always thought they were meant to seperate builtins, externals and internal packages

File: astropy/visualization/interval.py [hunk 5]
```diff
@@ -6,39 +6,73 @@
...
+        """
 
         vmin, vmax = self.get_limits(values)
 
         if out is None:
             values = np.subtract(values, float(vmin))
         else:
             if out.dtype.kind != 'f':
-                raise TypeError("Can only do in-place scaling for floating-point arrays")
+                raise TypeError('Can only do in-place scaling for '
```
Comment (MSeifert04) [inline astropy/visualization/interval.py]: Somwhere it was mentioned that error messages should start with a lowercase. Don't remember where though.

File: astropy/visualization/mpl_normalize.py [hunk 6]
```diff
@@ -1,14 +1,18 @@
 """
-Normalization class for Matplotlib that can be used to produce colorbars.
+Normalization class for Matplotlib that can be used to produce
+colorbars.
 """
 
 from __future__ import division, print_function
-
```
Comment (MSeifert04) [inline astropy/visualization/mpl_normalize.py]: same question here regarding the line removal.

File: astropy/visualization/mpl_normalize.py [hunk 7]
```diff
@@ -1,14 +1,18 @@
...
 import numpy as np
 from numpy import ma
+from .interval import (PercentileInterval, AsymmetricPercentileInterval,
+                       ManualInterval, MinMaxInterval)
+from .stretch import (LinearStretch, SqrtStretch, PowerStretch, LogStretch,
+                      AsinhStretch)
 
 try:
-    import matplotlib  # pylint: disable=W0611
+    import matplotlib    # pylint: disable=W0611
```
Comment (MSeifert04) [inline astropy/visualization/mpl_normalize.py]: why increase the number of whitespaces? two are recommended before inline comments.

File: astropy/visualization/stretch.py (hunk 1 above)
Comment (MSeifert04) [inline astropy/visualization/stretch.py]: should this be mentioned in the changelog?

File: astropy/visualization/stretch.py [hunk 8]
```diff
@@ -6,20 +6,19 @@
 """
 
 from __future__ import division, print_function
-
```
Comment (MSeifert04) [inline astropy/visualization/stretch.py]: same question here regarding the blank lines.

File: astropy/visualization/interval.py (hunk 4 above)
Comment (larrybradley) [inline astropy/visualization/interval.py]: I'm not a fan of the blank lines within the imports.  I wasn't aware that was recommended.

File: astropy/visualization/interval.py (hunk 5 above)
Comment (larrybradley) [inline astropy/visualization/interval.py]: I always write them as sentences (what if the message contains multiple sentences?).  In any case, @astrofrog actually wrote this one, I just wrapped the line.

File: astropy/visualization/interval.py (hunk 4 above)
Comment (MSeifert04) [inline astropy/visualization/interval.py]: I'm not sure if it's really recommended. [PEP8](https://www.python.org/dev/peps/pep-0008/#imports) states they should be grouped and I always interpreted this as blank lines. :smile:

File: astropy/visualization/mpl_normalize.py (hunk 7 above)
Comment (larrybradley) [inline astropy/visualization/mpl_normalize.py]: I think it's *at least* two spaces.  I think this was inadvertent, so I could change it back.

File: astropy/visualization/interval.py (hunk 5 above)
Comment (MSeifert04) [inline astropy/visualization/interval.py]: At least the python-builtin exceptions start lowercase. I think the rational was that the sentence starts with the capitalized exception name and then the first sentence naturally would be lowercase.

I don't mind the capitalized start so feel free to ignore this.

File: astropy/visualization/interval.py (hunk 4 above)
Comment (larrybradley) [inline astropy/visualization/interval.py]: I do order them, just without blank lines. :)  Ah, OK.  `pep8` has failed me all these years (it could at least check the standard library imports).

File: astropy/visualization/mpl_normalize.py (hunk 7 above)
Comment (MSeifert04) [inline astropy/visualization/mpl_normalize.py]: You are right it's at least 2 spaces.

File: astropy/visualization/interval.py (hunk 4 above)
Comment (pllim) [inline astropy/visualization/interval.py]: Blank line also makes it stands out, so easier to remove when no longer needed.

Comment (larrybradley) [inline astropy/visualization/interval.py]: @pllim But the `__future__` import always goes first, so it's hard to miss! 😄

File: astropy/visualization/interval.py [hunk 9]
```diff
@@ -90,9 +129,9 @@ class AsymmetricPercentileInterval(BaseInterval):
     upper_percentile : float
         The upper percentile above which to ignore pixels.
     n_samples : int, optional
-        Maximum number of values to use. If this is specified, and there are
-        more values in the dataset as this, then values are randomly sampled
-        from the array (with replacement)
+        Maximum number of values to use. If this is specified, and there
+        are more values in the dataset as this, then values are randomly
+        sampled from the array (with replacement)
```
Comment (MSeifert04) [inline astropy/visualization/interval.py]: Missing dot at the end of the sentence.

File: astropy/visualization/interval.py [hunk 10]
```diff
@@ -130,15 +168,17 @@ class PercentileInterval(AsymmetricPercentileInterval):
         The fraction of pixels to keep. The same fraction of pixels is
         eliminated from both ends.
     n_samples : int, optional
-        Maximum number of values to use. If this is specified, and there are
-        more values in the dataset as this, then values are randomly sampled
-        from the array (with replacement)
+        Maximum number of values to use. If this is specified, and there
+        are more values in the dataset as this, then values are randomly
+        sampled from the array (with replacement)
```
Comment (MSeifert04) [inline astropy/visualization/interval.py]: same here (missing `.`)

File: astropy/visualization/interval.py [hunk 11]
```diff
@@ -152,23 +192,23 @@ class ZScaleInterval(BaseInterval):
     image : array_like
         Input array.
     nsamples : int, optional
-        Number of points in array to sample for determining scaling factors.
-        Default to 1000.
+        Number of points in array to sample for determining scaling
+        factors.  Default to 1000.
```
Comment (MSeifert04) [inline astropy/visualization/interval.py]: I don't think `Default to ...` is right. Shouldn't it be `DefaultS to ...` or `Default IS ...`

File: astropy/visualization/interval.py [hunk 12]
```diff
@@ -152,23 +192,23 @@ class ZScaleInterval(BaseInterval):
...
     contrast : float, optional
-        Scaling factor (between 0 and 1) for determining min and max. Larger
-        values increase the difference between min and max values used for
-        display. Default to 0.25.
+        Scaling factor (between 0 and 1) for determining min and max.
+        Larger values increase the difference between min and max values
+        used for display. Default to 0.25.
```
Comment (MSeifert04) [inline astropy/visualization/interval.py]: same here (default to ...)

File: astropy/visualization/interval.py [hunk 13]
```diff
@@ -152,23 +192,23 @@ class ZScaleInterval(BaseInterval):
...
     max_reject : float, optional
-        If more than ``max_reject * npixels`` pixels are rejected, then the
-        returned values are the min and max of the data. Default to 0.5.
+        If more than ``max_reject * npixels`` pixels are rejected, then
+        the returned values are the min and max of the data. Default to
```
Comment (MSeifert04) [inline astropy/visualization/interval.py]: and here.

File: astropy/visualization/mpl_normalize.py [hunk 14]
```diff
@@ -18,10 +22,11 @@ class Normalize(Normalize, object):
 except ImportError:
     class Normalize(object):
         def __init__(self, *args, **kwargs):
-            raise ImportError("matplotlib is required in order to use this class")
+            raise ImportError('matplotlib is required in order to use this '
+                              'class')
```
Comment (MSeifert04) [inline astropy/visualization/mpl_normalize.py]: missing `.` at the end of the sentence.

File: astropy/visualization/mpl_normalize.py [hunk 15]
```diff
@@ -30,24 +35,48 @@ class ImageNormalize(Normalize):
 
     Parameters
     ----------
+    data : `~numpy.ndarray`, optional
```
Comment (MSeifert04) [inline astropy/visualization/mpl_normalize.py]: If it's optional what is the default?

Comment (MSeifert04) [inline astropy/visualization/mpl_normalize.py]: okay, I think I understand why giving a default value might be a problem here.

File: astropy/visualization/stretch.py (hunk 1 above)
Comment (larrybradley) [inline astropy/visualization/stretch.py]: I suspect that not many people were actually calling stretch instances directly on the data (with or without the `out` keyword), but I'm happy to mention it in the changelog.  @astrofrog?

File: astropy/visualization/mpl_normalize.py [hunk 16]
```diff
@@ -86,9 +114,108 @@ def __call__(self, values, clip=None):
...
+
+    This function is used by the
+    ``astropy.visualization.scripts.fits2bitmap`` script.
+
+    Parameters
+    ----------
+    data : `~numpy.ndarray`
+        The image array.
+
+    stretch : {{'linear', 'sqrt', 'power', log', 'asinh'}}, optional
```
Comment (MSeifert04) [inline astropy/visualization/mpl_normalize.py]: why two curly braces? I've never seen this way of giving allowed values.

File: astropy/visualization/mpl_normalize.py [hunk 17]
```diff
@@ -86,9 +114,108 @@ def __call__(self, values, clip=None):
...
+        'linear'.
+
+    power : float, optional
+        The power index for ``stretch='power'``.  The default is 1.0.
+
+    asinh_a : float, optional
+        For ``stretch='asinh'``, the value where the asinh curve
+        transitions from linear to logarithmic behavior, expressed as a
+        fraction of the normalized image.  Must be in the range between
+        0 and 1.
```
Comment (MSeifert04) [inline astropy/visualization/mpl_normalize.py]: Missing default value?

File: astropy/visualization/mpl_normalize.py [hunk 18]
```diff
@@ -86,9 +114,108 @@ def __call__(self, values, clip=None):
...
+
+    min_cut : float, optional
+        The pixel value of the minimum cut level.  Data values less than
+        ``min_cut`` will set to ``min_cut`` before stretching the image.
+        The default is the image minimum.  ``min_cut`` overrides
+        ``min_percent``.
```
Comment (MSeifert04) [inline astropy/visualization/mpl_normalize.py]: Missing default value?

File: astropy/visualization/mpl_normalize.py [hunk 19]
```diff
@@ -86,9 +114,108 @@ def __call__(self, values, clip=None):
...
+
+    max_cut : float, optional
+        The pixel value of the maximum cut level.  Data values greater
+        than ``min_cut`` will set to ``min_cut`` before stretching the
+        image.  The default is the image maximum.  ``max_cut`` overrides
+        ``max_percent``.
```
Comment (MSeifert04) [inline astropy/visualization/mpl_normalize.py]: here too

File: astropy/visualization/mpl_normalize.py (hunk 18 above)
Comment (MSeifert04) [inline astropy/visualization/mpl_normalize.py]: ups, I can't read. I just spotted it. Ignore this comment.

File: astropy/visualization/mpl_normalize.py [hunk 20]
```diff
@@ -86,9 +114,108 @@ def __call__(self, values, clip=None):
...
+    elif stretch == 'sqrt':
+        stretch = SqrtStretch()
+    elif stretch == 'power':
+        stretch = PowerStretch(power)
+    elif stretch == 'log':
+        stretch = LogStretch()
+    elif stretch == 'asinh':
+        stretch = AsinhStretch(asinh_a)
+    else:
+        raise ValueError('Unknown stretch: {0}'.format(stretch))
```
Comment (MSeifert04) [inline astropy/visualization/mpl_normalize.py]: missing `.` at the end of the sentence (and maybe the capitalization).

File: astropy/visualization/scripts/fits2bitmap.py [hunk 21]
```diff
@@ -28,25 +28,24 @@ def fits2bitmap(filename, ext=0, out_fn=None, scale='linear',
...
+        The power index for ``stretch='power'``.  The default is 1.0.
     asinh_a : float, optional
-        For ``scale='asinh'`` image scaling, the value where the asinh
-        curve transitions from linear to logarithmic behavior, expressed
-        as a fraction of the normalized image.  Must be in the range
-        between 0 and 1.
+        For ``stretch='asinh'``, the value where the asinh curve
+        transitions from linear to logarithmic behavior, expressed as a
+        fraction of the normalized image.  Must be in the range between
+        0 and 1.
```
Comment (MSeifert04) [inline astropy/visualization/scripts/fits2bitmap.py]: missing default value.

File: astropy/visualization/scripts/fits2bitmap.py [hunk 22]
```diff
@@ -95,16 +93,16 @@ def fits2bitmap(filename, ext=0, out_fn=None, scale='linear',
         out_fn += '.png'
 
     if cmap not in cm.datad:
-        log.critical('{0} is not a valid matplotlib colormap '
-                     'name'.format(cmap))
+        log.critical('{0} is not a valid matplotlib colormap name'
```
Comment (MSeifert04) [inline astropy/visualization/scripts/fits2bitmap.py]: missing `.` at the end of the sentence.

File: astropy/visualization/scripts/fits2bitmap.py [hunk 23]
```diff
@@ -95,16 +93,16 @@ def fits2bitmap(filename, ext=0, out_fn=None, scale='linear',
...
-                               max_cut=max_cut, min_percent=min_percent,
-                               max_percent=max_percent, percent=percent)
+    norm = simple_norm(image, stretch=stretch, power=power, asinh_a=asinh_a,
+                       min_cut=min_cut, max_cut=max_cut,
+                       min_percent=min_percent, max_percent=max_percent,
+                       percent=percent)
 
-    mimg.imsave(out_fn, image_scaled, cmap=cmap)
+    mimg.imsave(out_fn, norm(image), cmap=cmap)
     log.info('Saved file to {0}'.format(out_fn))
```
Comment (MSeifert04) [inline astropy/visualization/scripts/fits2bitmap.py]: missing `.` at the end of the info message.

File: astropy/visualization/stretch.py [hunk 24]
```diff
@@ -344,8 +379,11 @@ class HistEqStretch(BaseStretch):
 
     Parameters
     ----------
-    data : float
-        The data defining the equalization
+    data : array-like
+        The data defining the equalization.
+    values : array-like, optional
+        The input image values, which should already be normalized to
+        the [0:1] range.
```
Comment (MSeifert04) [inline astropy/visualization/stretch.py]: Default value?

File: astropy/visualization/stretch.py [hunk 25]
```diff
@@ -362,22 +400,28 @@ def __init__(self, data, values=None):
...
     """
     Inverse transformation for `~astropy.image.scaling.HistEqStretch`.
+
+    Parameters
+    ----------
+    data : array-like
+        The data defining the equalization.
+    values : array-like, optional
+        The input image values, which should already be normalized to
+        the [0:1] range.
```
Comment (MSeifert04) [inline astropy/visualization/stretch.py]: does it make sense to include the default value?

File: astropy/visualization/ui.py [hunk 26]
```diff
@@ -1,14 +1,18 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 
 import numpy as np
-
```
Comment (MSeifert04) [inline astropy/visualization/ui.py]: I would suggest including this blank line again.

**Comment by MSeifert04** [comment]
I'm sorry for the flood of stylistic comments. I don't feel strong about them, so feel free to ignore them.

File: astropy/visualization/stretch.py (hunk 25 above)
Comment (larrybradley) [inline astropy/visualization/stretch.py]: I'm not sure (I don't use the histogram-equalized stretches).

**Comment by larrybradley** [comment]
Thanks, @MSeifert04.  I've implemented everything (except explaining the hist-eq `values` default).

@astrofrog  *Now* this should be ready to go.

**Comment by astrofrog** [comment]
Thanks @larrybradley and sorry for the delay in merging!
